builder/builder.go

```diff
@@ -24,10 +25,49 @@ type Builder struct {
 func New(ownAddr *felt.Felt, bc *blockchain.Blockchain, builderVM vm.VM, log utils.Logger) *Builder {
 	return &Builder{
 		ownAddress: *ownAddr,
+		bc:         bc,
+		vm:         builderVM,
+	}
+}
 
-		bc: bc,
-		vm: builderVM,
+func (b *Builder) Receipt(fee *felt.Felt, feeUnit core.FeeUnit, txHash *felt.Felt, reverted bool,
+	trace *vm.TransactionTrace,
+) (*core.TransactionReceipt, error) {
+	executionResources := make([]*vm.ExecutionResources, 0)
+	events := make([]*core.Event, 0)
+	messages := make([]*core.L2ToL1Message, 0)
+	if trace.ConstructorInvocation != nil {
+		executionResources = append(executionResources, trace.ConstructorInvocation.ExecutionResources)
+		events = append(events, utils.Map(trace.ConstructorInvocation.Events, vm2core.AdaptOrderedEvent)...)
+		messages = append(messages, utils.Map(trace.ConstructorInvocation.Messages, vm2core.AdaptMsgToL1)...)
```

Comment:
Don't we need to recursively traverse the call stack and accumulate all the events and messages?

Reply:
and we should maintain the order of the events and messages